benches/comparison.rs

```diff
@@ -0,0 +1,129 @@
+#![cfg(not(tarpaulin))]
+
+use criterion::{black_box, criterion_group, Bencher, Criterion};
+use ringbuffer::{AllocRingBuffer, ConstGenericRingBuffer, RingBuffer};
+use std::collections::{LinkedList, VecDeque};
+use std::sync::mpsc::channel;
+
+const ITER: usize = 1024 * 16;
+const CAP: usize = 1024;
+
+fn std_chan(b: &mut Bencher) {
+    let (tx, rx) = channel();
+
+    b.iter(|| {
+        for i in 0..ITER {
+            let _ = tx.send(i);
+            black_box(());
+        }
+
+        for i in 0..ITER {
+            let res = rx.recv();
+            let _ = black_box(res);
+        }
+    });
+}
+
+fn vec(b: &mut Bencher) {
+    let mut vd = Vec::with_capacity(CAP);
+
+    b.iter(|| {
+        for i in 0..ITER {
+            let _ = vd.push(i);
+            black_box(());
+        }
+
+        for i in 0..ITER {
+            let res = vd.remove(0);
+            let _ = black_box(res);
+        }
+    });
+}
+
+fn vecdeque(b: &mut Bencher) {
+    let mut vd = VecDeque::with_capacity(CAP);
+
+    b.iter(|| {
+        for i in 0..ITER {
+            let _ = vd.push_back(i);
+            black_box(());
+        }
+        for i in 0..ITER {
+            let res = vd.pop_front();
+            let _ = black_box(res);
+        }
+    });
+}
+
+fn linked_list(b: &mut Bencher) {
+    let mut ll = LinkedList::new();
+
+    b.iter(|| {
+        for i in 0..ITER {
+            let _ = ll.push_back(i);
+            black_box(());
+        }
+
+        for i in 0..ITER {
+            let res = ll.pop_front();
+            let _ = black_box(res);
+        }
+    });
+}
+
+fn cg_rb(b: &mut Bencher) {
+    let mut rb = ConstGenericRingBuffer::<_, CAP>::new();
+
+    b.iter(|| {
+        for i in 0..ITER {
+            let _ = rb.push(i);
+            black_box(());
+        }
+        for i in 0..ITER {
```

Comment:
Doing this more than CAP amount of time sis just a no-op right?

Reply:
sort of, the buffer has to reject the items all the rest of the times which might take time just as our dropping of old items does.

Reply:
sorry that's for the heapless version. Here it will drop old items ITER-CAP times

Reply:
I'm just kinda confused why we are testing different things for different structures, shouldn't we test all under the use case of:

A buffer which can contain at max CAP items, removing old entries as new ones get pushed. Clearing out the buffer at the end.

Reply:
no other datastructure has that behavior. They're different datastructures, we cannot test them under the same circumstances

Reply:
Then the comparison is meaningless imo, it should be about using similar datastructures for the same purpose.

Reply:
I disagree. No benchmark can completely fairly compare these datastructures, but it does give a good overview of differences in order of magnitude. Of course a channel is not a ringbuffer, but pushing 2^14 items to a channel takes about 14 times longer than to a ringbuffer which deals with removing old items. Simply pushing 2^14 items to a vecdeque that has to deal with growing is about 3x slower than pushing to a ringbuffer.  Pushing to a heapless deque which drops items on insert is a few percent slower than pushing to a ringbuffer which deletes old items.

None of them are the same datastructure, but it does show the order of magnitude of differences you can expect when using these datastructures. Each datastructure is good at different things, but this is their normal performance when you use them as a queue.

In fact, I don't really want to modify these datastructures to function identically to ringbuffer. Then we don't measure how these datastructures normally perform as a queue for the purpose they were designed for. Instead we measure how well we can turn the datastructure into a ringbuffer.

Reply:
As long as we add a disclaimer/description of the benchmarks in the readme this is fine then I think